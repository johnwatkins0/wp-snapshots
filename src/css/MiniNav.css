.MiniNav {
	width: 100%;
	max-width: 768px;
	margin: 0 auto;
}

.MiniNav__grid {
	display: grid;
	grid-template-columns: repeat(auto-fit, minmax(100px, 1fr));
	grid-gap: 0.75rem;
	justify-content: center;
}

.MiniNavLink img {
	width: 100%;
	height: auto;
}
