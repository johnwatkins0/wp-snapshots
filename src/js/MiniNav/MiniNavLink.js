/* eslint react/no-danger: 0 */
import React from 'react';
import PropTypes from 'prop-types';
import { NavLink } from 'react-router-dom';

const MiniNavLink = ({
  title: { rendered: title },
  image: {
    media_details: {
      sizes: { medium: image },
    },
  },
  slug,
}) => (
  <NavLink
    to={`/snapshot/${slug}/`}
    href={`/snapshot/${slug}/`}
    className="MiniNavLink"
    activeClassName="MiniNavLink--current"
    style={{ backgroundImage: `url('${image}')` }}
  >
    <img src={image.source_url} height={image.height} width={image.width} alt={title} />
  </NavLink>
);

MiniNavLink.propTypes = {
  slug: PropTypes.string.isRequired,
  title: PropTypes.objectOf(PropTypes.string).isRequired,
  image: PropTypes.objectOf(PropTypes.any).isRequired,
};

export default MiniNavLink;
