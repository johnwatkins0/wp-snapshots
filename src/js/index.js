/* eslint react/no-danger: 0 */
import React, { Component } from 'react';
import { render } from 'react-dom';
import PropTypes from 'prop-types';
import { BrowserRouter, Route, Switch, Link } from 'react-router-dom';
import DocumentTitle from 'react-document-title';
import get from 'lodash.get';
import { parse } from 'qs';

import Archive from './Archive';
import Single from './Single';
import MiniNav from './MiniNav';

import { appDidMount } from './store/actions';
import effects from './store/effects';
import reducer from './store/reducer';
import { getPostBySlug, getPosts } from './store/selectors';

import { ROOT_SELECTOR } from './constants';
import { withScrollToTop } from './utils/withScrollToTop';
import { withReplaceEditLink } from './utils/withReplaceEditLink';

const ScrollToTopSingle = withScrollToTop(withReplaceEditLink(Single, 'post'));
const ScrollToTopArchive = withScrollToTop(withReplaceEditLink(Archive, 'category'));

const INITIAL_STATE = {
  posts: [],
};

export class App extends Component {
  static propTypes = {
    termId: PropTypes.string.isRequired,
    termName: PropTypes.string.isRequired,
    termSlug: PropTypes.string.isRequired,
    termDescription: PropTypes.string.isRequired,
    basenamePrefix: PropTypes.string,
    titleAppend: PropTypes.string,
  };

  static defaultProps = {
    basenamePrefix: '',
    titleAppend: '',
  };

  constructor(props) {
    super(props);

    this.state = INITIAL_STATE;
    this.dispatch = this.dispatch.bind(this);
    this.triggerEffect = this.triggerEffect.bind(this);
    this.getState = this.getState.bind(this);
  }

  componentDidMount() {
    this.dispatch(appDidMount(this.props.termId, this.props.previewPost || null));
  }

  getState = () => ({ ...this.state });

  dispatch(action) {
    this.setState(state => reducer(state, action), () => this.triggerEffect(action));
  }

  triggerEffect(action) {
    if (action.type in effects) {
      effects[action.type](action, {
        dispatch: this.dispatch,
        getState: this.getState,
      });
    }
  }

  render = () => (
    <BrowserRouter basename={this.props.basenamePrefix}>
      <main className="SnapshotsMain">
        <Switch>
          <Route
            path="/snapshots/:name/"
            render={() => (
              <DocumentTitle title={this.props.termName + this.props.titleAppend}>
                <ScrollToTopArchive
                  posts={getPosts(this.state)}
                  termName={this.props.termName}
                  termDescription={this.props.termDescription}
                  basename={this.props.basenamePrefix}
                  id={Number(this.props.termId)}
                />
              </DocumentTitle>
            )}
          />

          <Route
            path="/snapshot/:slug/"
            component={({
              match: {
                params: { slug },
              },
            }) => {
              const post = getPostBySlug(this.state, slug);
              return (
                <DocumentTitle
                  key={get(post, 'id', '')}
                  title={get(post, ['title', 'rendered'], document.title) + this.props.titleAppend}
                >
                  <div>
                    <div className="SnapshotsMain__title-container">
                      <h1>
                        <Link
                          href={`/snapshots/${this.props.termSlug}/`}
                          to={`/snapshots/${this.props.termSlug}/`}
                          dangerouslySetInnerHTML={{ __html: this.props.termName }}
                        />
                      </h1>
                    </div>
                    {post &&
                      post.image && (
                        <ScrollToTopSingle {...post} basename={this.props.basenamePrefix} />
                      )}
                    <MiniNav
                      posts={getPosts(this.state)}
                      termName={this.props.termName}
                      termSlug={this.props.termSlug}
                    />
                  </div>
                </DocumentTitle>
              );
            }}
          />

          <Route
            path="/"
            component={({ location }) => {
              const search = location.search || '';
              const params = parse(search.replace('?', ''));
              if (!('preview' in params) || !('p' in params)) {
                return null;
              }

              if (!this.props.previewPost) {
                return null;
              }

              const post = this.state.posts.find(({ id }) => id === Number(params.p));
              return (
                <DocumentTitle
                  key={get(post, 'id', '')}
                  title={get(post, ['title', 'rendered'], document.title) + this.props.titleAppend}
                >
                  <div>
                    <div className="SnapshotsMain__title-container">
                      <h1>
                        <Link
                          href={`/snapshots/${this.props.termSlug}/`}
                          to={`/snapshots/${this.props.termSlug}/`}
                          dangerouslySetInnerHTML={{ __html: this.props.termName }}
                        />
                      </h1>
                    </div>
                    {post &&
                      post.image && (
                        <ScrollToTopSingle {...post} basename={this.props.basenamePrefix} />
                      )}
                    <MiniNav
                      posts={getPosts(this.state)}
                      termName={this.props.termName}
                      termSlug={this.props.termSlug}
                    />
                  </div>
                </DocumentTitle>
              );
            }}
          />

        </Switch>
      </main>
    </BrowserRouter>
  );
}

export const startApp = () => {
  [...document.querySelectorAll(ROOT_SELECTOR)].forEach((root) => {
    render(
      <App
        termName={root.getAttribute('data-term-name')}
        titleAppend={root.getAttribute('data-title-append')}
        termId={root.getAttribute('data-term-id')}
        termSlug={root.getAttribute('data-term-slug')}
        termDescription={root.getAttribute('data-term-description')}
        basenamePrefix={root.getAttribute('data-basename-prefix').replace(/\/$/, '') || ''}
        previewPost={JSON.parse(root.getAttribute('data-preview-post'))}
      />,
      root,
    );
  });
};

export const start = async () => {
  if (!('fetch' in window)) {
    await import('whatwg-fetch');
  }

  startApp();
};

window.addEventListener('load', start);
