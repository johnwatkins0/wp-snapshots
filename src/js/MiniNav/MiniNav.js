/* eslint react/no-danger: 0 */
import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import MiniNavLink from './MiniNavLink';

const MiniNav = ({ posts, termName, termSlug }) => (
  <nav className="MiniNav">
    <h2>
      <Link
        href={`/snapshots/${termSlug}/`}
        to={`/snapshots/${termSlug}/`}
        dangerouslySetInnerHTML={{ __html: termName }}
      />
    </h2>
    <div className="MiniNav__grid">
      {posts.map(post => post.image && <MiniNavLink key={post.id} {...post} />)}
    </div>
  </nav>
);

MiniNav.propTypes = {
  posts: PropTypes.arrayOf(PropTypes.object).isRequired,
  termName: PropTypes.string.isRequired,
  termSlug: PropTypes.string.isRequired,
};

export default MiniNav;
