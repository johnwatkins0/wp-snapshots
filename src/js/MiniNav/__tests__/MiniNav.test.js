import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { BrowserRouter, Route } from 'react-router-dom';

import { testPostsWithMedia } from '../../../../demo/src/data';

import MiniNav from '..';

Enzyme.configure({ adapter: new Adapter() });

const Component = () => (
  <BrowserRouter>
    <Route>
      <MiniNav termName="Name" termSlug="name" posts={testPostsWithMedia} />
    </Route>
  </BrowserRouter>
);

describe('MiniNav', () => {
  it('renders', () => {
    const tree = renderer.create(<Component />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('has the correct number of items', () => {
    const wrapper = mount(<Component />);
    const navLinks = wrapper.find('.MiniNavLink img');
    expect(navLinks.length).toBe(7);
  });
});
