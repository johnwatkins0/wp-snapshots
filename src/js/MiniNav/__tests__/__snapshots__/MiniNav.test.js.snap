// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MiniNav renders 1`] = `
<nav
  className="MiniNav"
>
  <h2>
    <a
      dangerouslySetInnerHTML={
        Object {
          "__html": "Name",
        }
      }
      href="/snapshots/name/"
      onClick={[Function]}
    />
  </h2>
  <div
    className="MiniNav__grid"
  >
    <a
      aria-current={false}
      className="MiniNavLink"
      href="/snapshot/donec-vel-volutpat/"
      onClick={[Function]}
      style={
        Object {
          "backgroundImage": "url('[object Object]')",
        }
      }
    >
      <img
        alt="Donec vel volutpat"
        height={200}
        src="/demo/src/images/300x200-7.jpg"
        width={300}
      />
    </a>
    <a
      aria-current={false}
      className="MiniNavLink"
      href="/snapshot/donec-ut-eros/"
      onClick={[Function]}
      style={
        Object {
          "backgroundImage": "url('[object Object]')",
        }
      }
    >
      <img
        alt="Donec ut eros"
        height={200}
        src="/demo/src/images/300x200-6.jpg"
        width={300}
      />
    </a>
    <a
      aria-current={false}
      className="MiniNavLink"
      href="/snapshot/vivamus-ullamcorper/"
      onClick={[Function]}
      style={
        Object {
          "backgroundImage": "url('[object Object]')",
        }
      }
    >
      <img
        alt="Vivamus ullamcorper"
        height={200}
        src="/demo/src/images/300x200-5.jpg"
        width={300}
      />
    </a>
    <a
      aria-current={false}
      className="MiniNavLink"
      href="/snapshot/vestibulum-dapibus/"
      onClick={[Function]}
      style={
        Object {
          "backgroundImage": "url('[object Object]')",
        }
      }
    >
      <img
        alt="Vestibulum dapibus"
        height={200}
        src="/demo/src/images/300x200-4.jpg"
        width={300}
      />
    </a>
    <a
      aria-current={false}
      className="MiniNavLink"
      href="/snapshot/ut-eget-orci/"
      onClick={[Function]}
      style={
        Object {
          "backgroundImage": "url('[object Object]')",
        }
      }
    >
      <img
        alt="Ut eget orci"
        height={200}
        src="/demo/src/images/300x200-3.jpg"
        width={300}
      />
    </a>
    <a
      aria-current={false}
      className="MiniNavLink"
      href="/snapshot/praesent-sollicitudin/"
      onClick={[Function]}
      style={
        Object {
          "backgroundImage": "url('[object Object]')",
        }
      }
    >
      <img
        alt="Praesent sollicitudin"
        height={200}
        src="/demo/src/images/300x200-2.jpg"
        width={300}
      />
    </a>
    <a
      aria-current={false}
      className="MiniNavLink"
      href="/snapshot/nullam-ultricies/"
      onClick={[Function]}
      style={
        Object {
          "backgroundImage": "url('[object Object]')",
        }
      }
    >
      <img
        alt="Nullam ultricies"
        height={200}
        src="/demo/src/images/300x200-1.jpg"
        width={300}
      />
    </a>
  </div>
</nav>
`;
